packages/vite/src/node/plugins/html.ts

```diff
@@ -938,6 +940,8 @@ export interface IndexHtmlTransformContext {
   bundle?: OutputBundle
```

Comment:
Wanted to double check something... The type signature of `IndexHtmlTransformContext` currently permits some invalid values, such as objects with `server` _and_ `bundle`. I wonder, why this type not a disjunction?

```ts
interface IndexHtmlTransformBuildContext {
  path: string
  filename: string
  // NOTE: `bundle` and `chunk` are optional because they are not set for
  // `transformIndexHtml` with `order: pre`.
  bundle?: OutputBundle
  chunk?: OutputChunk
}

interface IndexHtmlTransformDevContext {
  path: string
  filename: string
  server: ViteDevServer
  originalUrl: string
  // NOTE: `req` and `res` are optional because they are optional in
  // `ViteDevServer.transformIndexHtml`
  req?: Connect.IncomingMessage
  res?: http.ServerResponse
}

export type IndexHtmlTransformDevContext =
  IndexHtmlTransformBuildContext | IndexHtmlTransformDevContext;
```

Is it because such a change would break types for upstream dependents?

Alternatively, we could keep the existing type signature but group related fields together under commented groups, like:

```ts
export interface IndexHtmlTransformContext {
  /**
   * public path when served
   */
  path: string
  /**
   * filename on disk
   */
  filename: string

  // Parameters available during `vite dev` only
  server?: ViteDevServer
  originalUrl?: string
  req?: Connect.IncomingMessage
  res?: http.ServerResponse

  // Parameters available during `vite build only`
  bundle?: OutputBundle
  chunk?: OutputChunk
}
```

Reply:
Probably was easier implemented in the past that way 😄 But I think we could change that separately, but yeah I think it would break downstream quite a bit, so maybe not quite worth the change.